src/group_theory/group_action/sub_mul_action.lean

```diff
@@ -35,6 +35,43 @@ variables {S : Type u'} {T : Type u''} {R : Type u} {M : Type v}
 
 set_option old_structure_cmd true
 
+/-- `smul_mem_class S R M` says `S` is a type of subsets `s ≤ M` that are closed under the
+scalar action of `R` on `M`. -/
+class smul_mem_class (S : Type*) (R M : out_param $ Type*) [has_smul R M] [set_like S M] :=
+(smul_mem : ∀ {s : S} (r : R) {m : M}, m ∈ s → r • m ∈ s)
+
+/-- `vadd_mem_class S R M` says `S` is a type of subsets `s ≤ M` that are closed under the
+additive action of `R` on `M`. -/
+class vadd_mem_class (S : Type*) (R M : out_param $ Type*) [has_vadd R M] [set_like S M] :=
+(vadd_mem : ∀ {s : S} (r : R) {m : M}, m ∈ s → r +ᵥ m ∈ s)
+
+attribute [to_additive] smul_mem_class
+
+namespace set_like
+
+variables [has_smul R M] [set_like S M] [hS : smul_mem_class S R M] (s : S)
+include hS
+
+open smul_mem_class
+
+/-- A subset closed under the scalar action inherits that action. -/
+@[to_additive "A subset closed under the additive action inherits that action.",
+priority 900] -- lower priority so other instances are found first
+instance has_smul : has_smul R s := ⟨λ r x, ⟨r • x.1, smul_mem r x.2⟩⟩
+
+@[simp, norm_cast, to_additive, priority 900]
+-- lower priority so later simp lemmas are used first; to appease simp_nf
+lemma coe_smul (r : R) (x : s) : (↑(r • x) : M) = r • x := rfl
```

Comment:
This is likely to conflict with the unnamespaced `coe_smul` in #17048 so if you're okay with the inconvenience, could you make this lemma `protected`? It would be a temporary inconvenience since we can go back again to writing only `coe_smul` after #17048.
```suggestion
protected lemma coe_smul (r : R) (x : s) : (↑(r • x) : M) = r • x := rfl
```